### DISCUSSION TITLE
How does "run" work / why doesn't my Doodle application draw it's output?

### DISCUSSION THREAD

**Discussion by noelwelsh** [scalameta/metals#5005] [Q&A]
My Doodle / Java2D applications don't render their output when they are run via the Metals code lens. They run just fine using the sbt `run` command. A simple example is at https://github.com/creativescala/creative-scala-template

I might be able to debug this if I understood how Metals runs a project. Is there an explanation somewhere I can read? My initial guess is that the Java2D / Swing code runs in a daemon thread and Metals is not waiting for that thread to finish before it exits the program.

**Comment by tgodzik** [comment]
Thanks for the question! It seems that the program ends and closes the window. It seems to work ok if I add `StdIn.readChar()` after `image.draw()` :thinking: 

This seems a correct behaviour though, since there is nothing for the program to wait for.

**Comment by noelwelsh** [comment]
Thanks. It must be the UI is running in a daemon thread, the existence of which doesn't stop the JVM from exiting.
